### PR TITLE
Fleet provisioning/release to v4 beta

### PR DESCRIPTION
**Pull request by aggarw13**
*Issue #, if available:*

*Description of changes:*
Release the (Fleet) Provisioning library (from `feat/csr-in-fleet-provisioning` branch) to `v4_beta` branch. 
Update `README.md` to contain information about the Provisioning library and AWS IoT setup for running demos

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -57,10 +58,28 @@ script:
...
+        echo -e $AWS_IOT_CLIENT_CERT > credentials/clientCert.pem;
+      fi
+    fi
+  - if [ "$TRAVIS_PULL_REQUEST" = "false" ]; then 
+      if [ "$RUN_TEST" == "provisioning" ]; then
+        echo -e $AWS_IOT_GAMMA_PRIVATE_KEY > credentials/privateKey.pem;
+      else
+        echo -e $AWS_IOT_PRIVATE_KEY > credentials/privateKey.pem;
+      fi
+    fi
```
Comment (leegeth) [inline .travis.yml:74]: can this 2 ifs combined?

File: .travis.yml [hunk 2]
```diff
@@ -57,10 +58,28 @@ script:
...
   # Set credential defines passed to CMake when not a pull request build.
-  - if [ "$TRAVIS_PULL_REQUEST" = "false" ]; then export AWS_IOT_CREDENTIAL_DEFINES="-DIOT_TEST_SERVER=\"\\\"$AWS_IOT_ENDPOINT\\\"\" -DIOT_TEST_PORT=443 -DIOT_TEST_ROOT_CA=\"\\\"../credentials/AmazonRootCA1.pem\\\"\" -DIOT_TEST_CLIENT_CERT=\"\\\"../credentials/clientCert.pem\\\"\" -DIOT_TEST_PRIVATE_KEY=\"\\\"../credentials/privateKey.pem\\\"\""; fi
+  - if [ "$TRAVIS_PULL_REQUEST" = "false" ]; then
+      if [ "$RUN_TEST" == "provisioning" ]; then 
+        export AWS_IOT_CREDENTIAL_DEFINES="-DIOT_TEST_SERVER=\"\\\"$AWS_IOT_GAMMA_ENDPOINT\\\"\" -DIOT_TEST_PORT=443 -DIOT_TEST_ROOT_CA=\"\\\"../credentials/AmazonRootCA1.pem\\\"\" -DIOT_TEST_CLIENT_CERT=\"\\\"../credentials/clientCert.pem\\\"\" -DIOT_TEST_PRIVATE_KEY=\"\\\"../credentials/privateKey.pem\\\"\"";
+      else
+        export AWS_IOT_CREDENTIAL_DEFINES="-DIOT_TEST_SERVER=\"\\\"$AWS_IOT_ENDPOINT\\\"\" -DIOT_TEST_PORT=443 -DIOT_TEST_ROOT_CA=\"\\\"../credentials/AmazonRootCA1.pem\\\"\" -DIOT_TEST_CLIENT_CERT=\"\\\"../credentials/clientCert.pem\\\"\" -DIOT_TEST_PRIVATE_KEY=\"\\\"../credentials/privateKey.pem\\\"\"";
+      fi
```
Comment (leegeth) [inline .travis.yml:81]: can this common code for IOT_TEST_ROOT_CA, DIOT_TEST_CLIENT_CERT and DIOT_TEST_PRIVATE_KEY be moved out of the if?

File: README.md [hunk 3]
```diff
@@ -37,39 +35,51 @@ This SDK builds with [CMake](https://cmake.org/), a cross-platform build tool. T
...
-    - Set `IOT_DEMO_ROOT_CA` to the path of the root CA certificate downloaded in [step 1.3](https://docs.aws.amazon.com/iot/latest/developerguide/create-device-certificate.html). The corresponding command line option for this constant is `-r`.
-    - Set `IOT_DEMO_CLIENT_CERT` to the path of the client certificate downloaded in [step 1.3](https://docs.aws.amazon.com/iot/latest/developerguide/create-device-certificate.html). The corresponding command line option for this constant is `-c`.
-    - Set `IOT_DEMO_PRIVATE_KEY` to the path of the private downloaded in [step 1.3](https://docs.aws.amazon.com/iot/latest/developerguide/create-device-certificate.html). The corresponding command line option for this constant is `-k`.
+    - Set `IOT_DEMO_ROOT_CA` to the path of the root CA certificate downloaded during the setup of provisioning claim in step 2 of [Build Steps](https://github.com/aws/aws-iot-device-sdk-embedded-C-staging/tree/fleetprovisioning_beta#build-steps). The corresponding command line option for this constant is `-r`.
```
Comment (leegeth) [inline README.md:66]: this is generic to all the libraries right. I think the correct reference to the step is AWS IoT Account Setup 1. Also the link is pointing to the branch fleetprovisioning_beta which is not right.

File: README.md [hunk 4]
```diff
@@ -37,39 +35,51 @@ This SDK builds with [CMake](https://cmake.org/), a cross-platform build tool. T
...
+    - Set `IOT_DEMO_CLIENT_CERT` to the path of the client certificate downloaded during the setup of provisioning claim in step 2 of [Build Steps](https://github.com/aws/aws-iot-device-sdk-embedded-C-staging/tree/fleetprovisioning_beta#build-steps). The corresponding command line option for this constant is `-c`.
```
Comment (leegeth) [inline README.md:67]: same as above

File: README.md [hunk 5]
```diff
@@ -37,39 +35,51 @@ This SDK builds with [CMake](https://cmake.org/), a cross-platform build tool. T
...
+    - Set `IOT_DEMO_PRIVATE_KEY` to the path of the private downloaded during the setup of provisioning claim in step 2 of [Build Steps](https://github.com/aws/aws-iot-device-sdk-embedded-C-staging/tree/fleetprovisioning_beta#build-steps). The corresponding command line option for this constant is `-k`.
```
Comment (leegeth) [inline README.md:68]: same as above

File: libraries/aws/defender/include/aws_iot_defender.h [hunk 6]
```diff
@@ -65,13 +65,19 @@
  * @brief Bit flags or metrics used by @ref defender_function_setmetrics function.
```
Comment (leegeth) [inline libraries/aws/defender/include/aws_iot_defender.h:65]: why is there any change in defender?

File: README.md (hunk 3 above)
Comment (aggarw13) [reply README.md:66]: Yes, it is generic to all libraries.  Good catch. Updated the text and link

File: README.md (hunk 4 above)
Comment (aggarw13) [reply README.md:67]: Fixed

File: README.md (hunk 5 above)
Comment (aggarw13) [reply README.md:68]: Fixed

File: .travis.yml (hunk 1 above)
Comment (aggarw13) [reply .travis.yml:74]: The nested ifs are only temporary for running Provisioning tests/demos against the Gamma endpoint. I will create a separate PR to remove these

File: .travis.yml (hunk 2 above)
Comment (aggarw13) [reply .travis.yml:81]: I had tried commonizing it earlier but had run into TravisCI issues. Anyway, these nested ifs are temporary, and will be removed in a separate PR tomorrow

File: libraries/aws/defender/include/aws_iot_defender.h (hunk 6 above)
Comment (aggarw13) [reply libraries/aws/defender/include/aws_iot_defender.h]: May have been an artifact from an older `v4_beta` sync. Good catch. Overwrote with the latest version from `v4_beta`

File: .travis.yml (hunk 1 above)
Comment (leegeth) [reply .travis.yml:74]: Ok

File: .travis.yml (hunk 2 above)
Comment (leegeth) [reply .travis.yml:81]: ok

File: README.md [hunk 7]
```diff
@@ -37,39 +35,51 @@ This SDK builds with [CMake](https://cmake.org/), a cross-platform build tool. T
...
-    - Set `IOT_DEMO_ROOT_CA` to the path of the root CA certificate downloaded in [step 1.3](https://docs.aws.amazon.com/iot/latest/developerguide/create-device-certificate.html). The corresponding command line option for this constant is `-r`.
-    - Set `IOT_DEMO_CLIENT_CERT` to the path of the client certificate downloaded in [step 1.3](https://docs.aws.amazon.com/iot/latest/developerguide/create-device-certificate.html). The corresponding command line option for this constant is `-c`.
-    - Set `IOT_DEMO_PRIVATE_KEY` to the path of the private downloaded in [step 1.3](https://docs.aws.amazon.com/iot/latest/developerguide/create-device-certificate.html). The corresponding command line option for this constant is `-k`.
+    - Set `IOT_DEMO_ROOT_CA` to the path of the root CA certificate downloaded when setting up device certificate (or Provisioning Claim for Fleet Provisioning) in [AWS IoT Account Setup](https://github.com/aws/aws-iot-device-sdk-embedded-C-staging/tree/v4_beta#aws-iot-account-setup). The corresponding command line option for this constant is `-r`.
```
Comment (leegeth) [inline README.md:66]: https://github.com/aws/aws-iot-device-sdk-embedded-C-staging/tree/v4_beta#aws-iot-account-setup this is staging repo link

File: README.md [hunk 8]
```diff
@@ -37,39 +35,51 @@ This SDK builds with [CMake](https://cmake.org/), a cross-platform build tool. T
...
+    - Set `IOT_DEMO_CLIENT_CERT` to the path of the client certificate downloaded when setting up device certificate (or Provisioning Claim for Fleet Provisioning) in [AWS IoT Account Setup](https://github.com/aws/aws-iot-device-sdk-embedded-C-staging/tree/v4_beta#aws-iot-account-setup). The corresponding command line option for this constant is `-c`.
```
Comment (leegeth) [inline README.md:67]: staging repo link

File: README.md [hunk 9]
```diff
@@ -37,39 +35,51 @@ This SDK builds with [CMake](https://cmake.org/), a cross-platform build tool. T
...
+    - Set `IOT_DEMO_PRIVATE_KEY` to the path of the private downloaded when setting up device certificate (or Provisioning Claim for Fleet Provisioning) in [AWS IoT Account Setup](https://github.com/aws/aws-iot-device-sdk-embedded-C-staging/tree/v4_beta#aws-iot-account-setup). The corresponding command line option for this constant is `-k`.
```
Comment (leegeth) [inline README.md:68]: same as above

File: README.md (hunk 7 above)
Comment (aggarw13) [reply README.md:66]: Another good catch. Seems like this was a bug from the beginning.

File: README.md (hunk 9 above)
Comment (aggarw13) [reply README.md]: Fixed

File: README.md (hunk 8 above)
Comment (aggarw13) [reply README.md]: Fixed

File: README.md (hunk 7 above)
Comment (aggarw13) [reply README.md]: Fixed

File: README.md [hunk 10]
```diff
@@ -19,13 +19,11 @@ This library is a new design that inherits from both the AWS IoT Device SDK Embe
...
+- Device Defender client library for the [AWS IoT Device Defender](https://docs.aws.amazon.com/iot/latest/developerguide/device-defender.html) service.
+- Provisioning client library for the [Fleet Provisioning  feature
+of AWS IoT Core](https://docs.aws.amazon.com/iot/latest/developerguide/provision-wo-cert.html).
 
-Compared to the AWS IoT Device SDK Embedded C v3.0.1, the following features were removed:
-- Auto-reconnect for MQTT connections.
-- Shadow JSON document generator.
-
-## Building and Running Demos
+## Building and Running the Fleet Provisioning library Demo
```
Comment (aggarg) [inline README.md:26]: Is this section specific to Fleet provisioning demo? What about other demos?

File: README.md [hunk 11]
```diff
@@ -37,39 +35,51 @@ This SDK builds with [CMake](https://cmake.org/), a cross-platform build tool. T
...
+    - Set `IOT_DEMO_CLIENT_CERT` to the path of the client certificate downloaded when setting up device certificate (or Provisioning Claim for Fleet Provisioning) in [AWS IoT Account Setup](https://github.com/aws/aws-iot-device-sdk-embedded-C/tree/v4_beta#aws-iot-account-setup). The corresponding command line option for this constant is `-c`.
+    - Set `IOT_DEMO_PRIVATE_KEY` to the path of the private downloaded when setting up device certificate (or Provisioning Claim for Fleet Provisioning) in [AWS IoT Account Setup](https://github.com/aws/aws-iot-device-sdk-embedded-C/tree/v4_beta#aws-iot-account-setup). The corresponding command line option for this constant is `-k`.
     - Set `IOT_DEMO_USER_NAME` to the username string, if any, required to authenticate to your MQTT broker. The corresponding command line option for this constant is `-m`.
     - Set `IOT_DEMO_PASSWORD` to the password string, if any, required to authenticate to your MQTT broker. The corresponding command line option for this constant is `-w`.
-4. Make a build directory in the SDK's root directory and `cd` into it.
+5. Make a build directory in the SDK's root directory and `cd` into it.
```
Comment (aggarg) [inline README.md:71]: `5` after `3`?

File: demos/iot_config.h [hunk 12]
```diff
@@ -55,30 +55,64 @@
...
+ * for registering thing in the Provisioning demo applications.
+ */
+#define AWS_IOT_DEMO_PROVISIONING_TEMPLATE_NAME    ""
+
+/**
+ * List of parameters that will be used for provisioning in the demo application.
+ * There are 2 requirements for passing parameters to the Provisioning demo:
+ * 1. One parameter is a "SerialNumber". The AWS IoT console generates a provisioning template
+ * that by default uses a "SerialNumber" device parameter for creating a Thing resource. Also, the
+ * demo setup instructions (on <PLACEHOLDER FOR LINK>) provide an AWS CLI command example for creating
```
Comment (aggarg) [inline demos/iot_config.h:77]: <PLACEHOLDER FOR LINK> - Is it ready?

File: demos/src/Provisioning/aws_iot_demo_provisioning_with_csr.c [hunk 13]
```diff
@@ -0,0 +1,594 @@
...
+                    pResponseInfo->u.acceptedResponse.pDeviceCert,
+                    pResponseInfo->u.acceptedResponse.certIdLength,
+                    pResponseInfo->u.acceptedResponse.pCertId,
+                    pResponseInfo->u.acceptedResponse.ownershipTokenLength,
+                    pResponseInfo->u.acceptedResponse.pCertOwnershipToken );
+    }
+    else
+    {
+        IotLogWarn( "demo: Oops, certificate creation from CSR has been rejected. "
+                    "Consider re-running the demo!" );
```
Comment (aggarg) [inline demos/src/Provisioning/aws_iot_demo_provisioning_with_csr.c:184]: "Consider re-running the demo!" --> Does not sound like a good message and I see that at multiple places.

File: tests/iot_config.h [hunk 14]
```diff
@@ -223,13 +275,13 @@
      * printed. */
     #define IotLogging_Puts       _coveragePuts
 
-    /* Includes for coverage logging puts. */
+/* Includes for coverage logging puts. */
```
Comment (aggarg) [inline tests/iot_config.h:278]: This indentation change seems off.

File: tests/iot_config.h [hunk 15]
```diff
@@ -223,13 +275,13 @@
...
     #include <stdbool.h>
     #include <stdio.h>
     #include <string.h>
 
-    /* Logging output function that only prints messages from demo executables.
-     * May be unused, hence the gcc unused attribute (not portable!) */
+/* Logging output function that only prints messages from demo executables.
```
Comment (aggarg) [inline tests/iot_config.h:283]: Same as above.

File: libraries/standard/serializer/src/cbor/iot_serializer_tinycbor_decoder.c [hunk 16]
```diff
@@ -296,9 +315,28 @@ static IotSerializerError_t _createDecoderObject( _cborValueWrapper_t * pCborVal
...
-                                            ( pDecoderObject->u.value.u.string.length ) );
+
+                        /* There is a special case for the last/ending binary/text string element
+                         * in an indefinite length container as indefinite length containers end with the
+                         * "break" byte (0xFF) to indicate the end of the container.
+                         * (Refer to the specification for "Indefinite-Length Arrays and Maps"
+                         * https://tools.ietf.org/html/rfc7049#section-2.2.2 )
+                         * Thus, when calculating the starting string location for this special case,
+                         * we have to account for the byte. */
+                        uint8_t * pPossibleBreakByte = ( uint8_t * ) ( next.ptr - 1 );
```
Comment (aggarg) [inline libraries/standard/serializer/src/cbor/iot_serializer_tinycbor_decoder.c:326]: This is an issue we are working on fixing and this does not seem to be the right patch!

File: README.md (hunk 10 above)
Comment (aggarw13) [reply README.md:26]: Oops, it is an artifact of the `FleetProvisioning_beta` README. Will remove the reference to the library from the section

File: tests/iot_config.h (hunk 14 above)
Comment (aggarw13) [reply tests/iot_config.h:278]: This is the formatting produced by Uncrustify-0.68-2-8c80bd84 on Linux..

File: demos/iot_config.h (hunk 12 above)
Comment (aggarw13) [reply demos/iot_config.h]: Good catch. Added the link

File: demos/src/Provisioning/aws_iot_demo_provisioning_with_csr.c (hunk 13 above)
Comment (aggarw13) [reply demos/src/Provisioning/aws_iot_demo_provisioning_with_csr.c]: Have updated the log messages to be helpful by suggesting to verify the attached IoT Policy for the demo credentials

File: README.md (hunk 11 above)
Comment (aggarw13) [reply README.md:71]: Fixed

File: libraries/standard/serializer/src/cbor/iot_serializer_tinycbor_decoder.c (hunk 16 above)
Comment (aggarw13) [reply libraries/standard/serializer/src/cbor/iot_serializer_tinycbor_decoder.c:326]: This patch (to our wrapper of the tinycbor decoder) was merged last year (https://github.com/aws/aws-iot-device-sdk-embedded-C-staging/pull/7) to support parsing of indefinite length containers by the Provisioning library

**Pull request merged**